components/client/src/main/java/com/hotels/styx/client/netty/connectionpool/FlowControllingHttpContentProducer.java

```diff
@@ -88,23 +87,23 @@
                 .transition(BUFFERING, ContentChunkEvent.class, this::contentChunkEventWhileBuffering)
                 .transition(BUFFERING, ChannelInactiveEvent.class, this::releaseAndTerminate)
                 .transition(BUFFERING, ChannelExceptionEvent.class, this::releaseAndTerminate)
-                .transition(BUFFERING, ChannelIdleEvent.class, this::releaseAndTerminate)
+                .transition(BUFFERING, DelayedTearDownEvent.class, this::releaseAndTerminate)
                 .transition(BUFFERING, ContentSubscribedEvent.class, this::contentSubscribedEventWhileBuffering)
                 .transition(BUFFERING, ContentEndEvent.class, this::contentEndEventWhileBuffering)
 
                 .transition(BUFFERING_COMPLETED, RxBackpressureRequestEvent.class, this::rxBackpressureRequestBeforeSubscription)
                 .transition(BUFFERING_COMPLETED, ContentChunkEvent.class, this::spuriousContentChunkEvent)
                 .transition(BUFFERING_COMPLETED, ChannelInactiveEvent.class, s -> BUFFERING_COMPLETED)
                 .transition(BUFFERING_COMPLETED, ChannelExceptionEvent.class, s -> BUFFERING_COMPLETED)
-                .transition(BUFFERING_COMPLETED, ChannelIdleEvent.class, this::releaseAndTerminate)
+                .transition(BUFFERING_COMPLETED, DelayedTearDownEvent.class, this::releaseAndTerminate)
                 .transition(BUFFERING_COMPLETED, ContentSubscribedEvent.class, this::contentSubscribedEventWhileBufferingCompleted)
                 .transition(BUFFERING_COMPLETED, ContentEndEvent.class, s -> BUFFERING_COMPLETED)
 
                 .transition(STREAMING, RxBackpressureRequestEvent.class, this::rxBackpressureRequestEventWhileStreaming)
                 .transition(STREAMING, ContentChunkEvent.class, this::contentChunkEventWhileStreaming)
                 .transition(STREAMING, ChannelInactiveEvent.class, e -> emitErrorAndTerminate(e.cause()))
                 .transition(STREAMING, ChannelExceptionEvent.class, e -> emitErrorAndTerminate(e.cause()))
-                .transition(STREAMING, ChannelIdleEvent.class, e -> emitErrorAndTerminate(new ResponseTimeoutException(origin)))
+                .transition(STREAMING, DelayedTearDownEvent.class, e -> emitErrorAndTerminate(e.cause()))
```

Comment:
I would use a different method for the state transition handler. For example:

```
+                .transition(STREAMING, DelayedTearDownEvent.class, e -> tearDown())
```

And then in the `tearDown` method:

```
private void tearDown() {
   emitErrorAndTerminate(new TimedOutExceptionWhatever());  // Choose an appropriate exception here.
}
```

So I would have a dedicated method specifically for the `tearDown` transition. This exposes the state transition logic more explicitly.

Reply:
That sounds like an improvement, but in your example we are emitting an error, I assume, to signal an invalid combination of state+event.  Doesn't it make more sense to use a new 'tearDown' method for the cases in which we are calling 'releaseAndTerminate' ?